### PR TITLE
add Change Password test

### PR DESCRIPTION
**Pull request by maivuhoa**

### CODE REVIEW

File: src/RailWay/java/Common/Utilities.java [hunk 1]
```diff
@@ -18,13 +18,13 @@ public static String generateRandomEmail(int len){
     }
 
     public static String generateRandomString(int len){
-        String email = RandomStringUtils.randomAlphanumeric(len);
-        return email;
+        String randomStr = RandomStringUtils.randomAlphanumeric(len);
+        return randomStr;
```
Comment (MinhQuan119) [inline src/RailWay/java/Common/Utilities.java]: this method can be shorter, please research and update

File: src/RailWay/java/PageObjects/ChangePassWordPage.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    private final By txtNewPassWord=By.xpath("//li[@class='new-password']/input[@id='newPassword']");
+    private final By txtConfirmPassWord=By.xpath("//li[@class='confirm-password']/input[@id='confirmPassword']");
+    private final By btnChangePassWord=By.xpath("//p[@class='form-actions']/input[@type='submit']");
+
+    public WebElement getTabChangePassWord(){return Constant.WEBDRIVER.findElement(tabChangePassWord);}
+    public WebElement getTxtCurrentPassWord(){return Constant.WEBDRIVER.findElement(txtCurrentPassWord);}
+    public WebElement getTxtNewPassWord(){return Constant.WEBDRIVER.findElement(txtNewPassWord);}
+    public WebElement getTxtConfirmPassWord(){return Constant.WEBDRIVER.findElement(txtConfirmPassWord);}
+    public WebElement getBtnChangePassWord(){return Constant.WEBDRIVER.findElement(btnChangePassWord);}
+    public void gotoTabChangePassWord(){this.getTabChangePassWord().click();}
```
Comment (MinhQuan119) [inline src/RailWay/java/PageObjects/ChangePassWordPage.java]: Please split this class into 3 parts as Powerpoint doc.
Besides, between two methods should have a blank row as what you did for `clickChangePassWordButton` and `changePass`

File: src/RailWay/java/PageObjects/ChangePassWordPage.java [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    public void clickChangePassWordButton(){this.getBtnChangePassWord();}
```
Comment (MinhQuan119) [inline src/RailWay/java/PageObjects/ChangePassWordPage.java]: Can it work without `this` ?

File: src/RailWay/java/PageObjects/ChangePassWordPage.java [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+    public void changePass(String curPass, String newPass, String confirmPass){
```
Comment (MinhQuan119) [inline src/RailWay/java/PageObjects/ChangePassWordPage.java]: Why did you use confirmPass but curPass?
Please update curPass to currentPass

File: src/RailWay/java/PageObjects/GeneralPage.java [hunk 5]
```diff
@@ -1,62 +1,62 @@
...
-    private final By txtWelComeMes=By.xpath("//div[@class='account']/strong");
+    private final By tabHome = By.xpath("//a[@href='../']");
+    private final By tabFAQ = By.xpath("//a[@href='/Page/FAQ.cshtml']");
+    private final By tabContact = By.xpath("//a[@href='/Page/Contact.cshtml']");
+    private final By tabTimeTable = By.xpath("//a[@href='TrainTimeListPage.cshtml']");
+    private final By tabTicketPrice = By.xpath("//a[@href='/Page/TrainPriceListPage.cshtml']");
+    private final By tabBookTicket = By.xpath("//a[@href='/Page/BookTicketPage.cshtml']");
+    private final By tabRegister = By.xpath("//a[@href='/Account/Register.cshtml']");
+    private final By tabLogin = By.xpath("//a[@href='/Account/Login.cshtml']");
+    private final By txtWelComeMes = By.xpath("//div[@class='account']/strong");
```
Comment (MinhQuan119) [inline src/RailWay/java/PageObjects/GeneralPage.java]: `txt` stands for textbox, you can't input on this message so it is not a textbox, it is a label (`lbl`)

File: src/RailWay/java/PageObjects/HomePage.java [hunk 6]
```diff
@@ -1,9 +1,10 @@
...
 
-public class HomePage extends GeneralPage{
-   public void  open(){
-       CONSTANT.WEBDRIVER.navigate().to(CONSTANT.RailwayURL);
-   }
+import Common.Constant;
+
+public class HomePage extends GeneralPage {
+    public void open() {
+        Constant.WEBDRIVER.navigate().to(Constant.RAILWAY_URL);
```
Comment (MinhQuan119) [inline src/RailWay/java/PageObjects/HomePage.java]: this method to navigate browser from a blank page to home page. Do not put it here

File: src/RailWay/java/PageObjects/LoginPage.java [hunk 7]
```diff
@@ -14,35 +14,41 @@ public class LoginPage extends GeneralPage {
     private final By stWelcomeText = By.xpath("//div[@class='account']/strong");
 
     private final By paErorrMessageLogin = By.xpath("//div[@id='content']/p[contains(@class,'error LoginForm')]");
```
Comment (MinhQuan119) [inline src/RailWay/java/PageObjects/LoginPage.java]: What do st and pa stands for?

**Review by MinhQuan119 — CHANGES_REQUESTED** [review]
It still has some more need to update. But, I think you should change those things first

File: .idea/.name [hunk 8]
```diff
@@ -0,0 +1 @@
+LoginTest.java
```
Comment (MinhQuan119) [inline .idea/.name:1]: files inside .idea folder is auto-generate files by building, please add them in git ignore and do not push it

File: pom.xml [hunk 9]
```diff
@@ -40,5 +40,15 @@
...
+            <artifactId>log4j-core</artifactId>
+            <version>2.17.2</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.logging.log4j</groupId>
+            <artifactId>log4j-api</artifactId>
+            <version>2.17.2</version>
+        </dependency>
     </dependencies>
 </project>
```
Comment (MinhQuan119) [inline pom.xml]: Missing ending line

File: src/RailWay/java/Common/constant/Constant.java [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package Common.constant;
+
+import org.openqa.selenium.WebDriver;
+
+public class Constant {
+    public static WebDriver WEBDRIVER;
+    public static final String RAILWAY_URL = "http://www.railwayb2.somee.com/Page/HomePage.cshtml";
+    public static final String LOGIN_USERNAME = "ccc@gmail.com";
+    public static final String LOGIN_BLANK_USERNAME = "";
```
Comment (MinhQuan119) [inline src/RailWay/java/Common/constant/Constant.java]: Do we need constants for blank? =))))

File: src/RailWay/java/PageObjects/ChangePassWordPage.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+import org.openqa.selenium.WebElement;
+
+public class ChangePassWordPage extends GeneralPage {
+    //Locator
+    private final By tabChangePassWord = By.xpath("//a[@href='/Account/ChangePassword.cshtml']");
+    private final By txtCurrentPassWord = By.xpath("//li[@class='current-password']/input[@id='currentPassword']");
+    private final By txtNewPassWord = By.xpath("//li[@class='new-password']/input[@id='newPassword']");
+    private final By txtConfirmPassWord = By.xpath("//li[@class='confirm-password']/input[@id='confirmPassword']");
+    private final By btnChangePassWord = By.xpath("//p[@class='form-actions']/input[@type='submit']");
+    private final By paSuccessChangePass = By.xpath("//form[@id='ChangePW']//p[@class='message success']");
```
Comment (MinhQuan119) [inline src/RailWay/java/PageObjects/ChangePassWordPage.java]: What does pa stand for? Please update refer this link
https://stackoverflow.com/questions/45028747/suggested-naming-conventions-for-selenium-identifiers

File: src/RailWay/java/PageObjects/ChangePassWordPage.java [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    private final By h1ChangePassword = By.xpath("//h1[text()='Change password']");
```
Comment (MinhQuan119) [inline src/RailWay/java/PageObjects/ChangePassWordPage.java]: Also for this one, it is `lbl`, not `h1`. `h1` is just a tag name on html

**Review by MinhQuan119 — CHANGES_REQUESTED** [review]
Please update

File: pom.xml (hunk 9 above)
Comment (maivuhoa) [reply pom.xml]: i have updated for it

File: src/RailWay/java/Common/constant/Constant.java (hunk 10 above)
Comment (maivuhoa) [reply src/RailWay/java/Common/constant/Constant.java]: i have removed it

File: src/RailWay/java/PageObjects/ChangePassWordPage.java (hunk 11 above)
Comment (maivuhoa) [reply src/RailWay/java/PageObjects/ChangePassWordPage.java]: i have update from pa to lbl

File: src/RailWay/java/PageObjects/ChangePassWordPage.java (hunk 12 above)
Comment (maivuhoa) [reply src/RailWay/java/PageObjects/ChangePassWordPage.java]: i have updated for it

File: src/RailWay/java/PageObjects/GeneralPage.java (hunk 5 above)
Comment (maivuhoa) [reply src/RailWay/java/PageObjects/GeneralPage.java]: i have updated for it

File: src/RailWay/java/PageObjects/LoginPage.java (hunk 7 above)
Comment (maivuhoa) [reply src/RailWay/java/PageObjects/LoginPage.java]: st is stands for Strong tag. I have updated from stWelcomeText to lblWelComeMessage.
pa is stands for p tag. I have updated from paErorrMessageLogin to msgErrorMessageLogin

File: src/RailWay/java/PageObjects/ChangePassWordPage.java (hunk 2 above)
Comment (maivuhoa) [reply src/RailWay/java/PageObjects/ChangePassWordPage.java]: I have update for it

File: src/RailWay/java/PageObjects/ChangePassWordPage.java (hunk 3 above)
Comment (maivuhoa) [reply src/RailWay/java/PageObjects/ChangePassWordPage.java]: Yes , it's can work without this

File: src/RailWay/java/PageObjects/ChangePassWordPage.java (hunk 4 above)
Comment (maivuhoa) [reply src/RailWay/java/PageObjects/ChangePassWordPage.java]: i have updated for it

File: src/RailWay/java/Common/Utilities.java (hunk 1 above)
Comment (maivuhoa) [reply src/RailWay/java/Common/Utilities.java]: i have updated for it

File: src/RailWay/java/PageObjects/HomePage.java (hunk 6 above)
Comment (maivuhoa) [reply src/RailWay/java/PageObjects/HomePage.java]: i have updated for it

File: .idea/.name (hunk 8 above)
Comment (maivuhoa) [reply .idea/.name:1]: i have updated for it

**Comment by maivuhoa** [comment]
Thanks your for reviewed !
I have updated for all issues.

**Review by MinhQuan119 — APPROVED** [review]
Approve để merge trước, vì nhiều chỗ cần sửa, để lâu sẽ gây dồn cục
Sau khi merge Hòa update lại code theo doc bên dưới, rồi tạo PR để check lại
https://docs.google.com/document/d/1OajWuBHSZc7jnPB7fyDvZx49Kl7e6B-JhbAPL6ZAZwM/edit

Doc chưa review phần test case, cái đó mình sẽ review tiếp sau khi merge r

**Pull request merged**